File: zd_02.cpp
#include <iostream>
#include <cmath>
#include <limits>

const double g = 9.81;  


double f(double alpha, double V, double L) {
    return (V * V * sin(2 * alpha)) / g - L;
}


double df(double alpha, double V) {
    return (2 * V * V * cos(2 * alpha)) / g;
}


double newton_method(double V, double L, double initial_guess, double tol = 1e-6, int max_iter = 100) {
    double alpha = initial_guess;
    for (int i = 0; i < max_iter; ++i) {
        double f_val = f(alpha, V, L);
        double df_val = df(alpha, V);
        
        // Avoid division by near-zero
        if (abs(df_val) < 1e-10) break;
        
        double delta = f_val / df_val;
        alpha -= delta;
        
        if (abs(delta) < tol) break;
    }
    return alpha;
}

int main() {
    double V, L;
    std::cout << "Enter initial velocity (V): ";
    std::cin >> V;
    std::cout << "Enter desired range (L): ";
    std::cin >> L;
    
    const double L_max = V * V / g;
    if (L > L_max) {
        std::cout << "Error: Desired range " << L 
                  << " m exceeds maximum possible range " << L_max << " m.\n";
    } else {

        double alpha_rad = newton_method(V, L, M_PI/4);
        double alpha_deg = alpha_rad * 180 / M_PI;
        
        // Validate physical angle bounds
        if (alpha_deg >= 0 && alpha_deg <= 90) {
            std::cout << "Required launch angle α: " << alpha_deg << "°\n";
        } else {
            std::cout << "Newton's Method failed to converge. Try another method.\n";
        }
    }
    return 0;
}

File: zd_02.py
import math

g = 9.81

V = float(input("Enter initial velocity (m/s): "))
L = float(input("Enter target distance (m): "))

max_L = V**2 / g
if L > max_L:
    print(f"Impossible! Maximum distance is {max_L:.2f} m")
else:

    if L < max_L/2:
        alpha = math.radians(30)
    else:
        alpha = math.radians(60)
    
    converged = False
    for _ in range(100):
        current = (V**2 * math.sin(2*alpha))/g - L
        slope = (2 * V**2 * math.cos(2*alpha))/g
        
        if abs(slope) < 1e-10:
            alpha += 0.1
            continue
            
        alpha -= current/slope
        
        if abs(current) < 1e-6:
            converged = True
            break
    
    angle_deg = math.degrees(alpha)
    
    if converged and 0 <= angle_deg <= 90:
        print(f"Launch angle needed: {angle_deg:.2f}°")
    else:
        print("No valid solution found (numerical error)")